src/reduced_basis/rb_parametrized_function.C

```diff
@@ -498,6 +515,33 @@ void RBParametrizedFunction::preevaluate_parametrized_function_on_mesh(const RBP
                   v.phi_i_all_qp[counter][i][j] = phi[i][j];
               }
             }
+            if (requires_all_elem_center_data)
+              {
+                v.qrule_orders[counter] = con.get_element_qrule().get_order();
+                // We try to minimize the number of reinit so we only do it once per elem for
+                // center quantities.
+                if (!elem_center_quantities_set)
+                  {
+                    // Get data derivatives at vertex average
+                    std::vector<Point> nodes = { elem_ref.reference_elem()->vertex_average() };
+                    elem_fe->reinit (&elem_ref, &nodes);
+
+                    // Here we do an implicit conversion from RealGradient which is a VectorValue<Real>
+                    // which in turn is a TypeVector<T> to a Point which is a TypeVector<Real>.
+                    // They are essentially the same thing. This helps us limiting the number of includes
+                    // in serialization and deserialization as RealGradient is a typedef and we cannot
+                    // forward declare typedefs. As a result we leverage the fact that point.h is already
+                    // includes in most places we need RealGradient.
+                    dxyzdxi_buffer = dxyzdxi[0];
+                    dxyzdeta_buffer = dxyzdeta[0];
+
+                    elem_center_quantities_set = true;
+                    elem_fe->reinit(&elem_ref);
```

Comment:
Can you check the new version ? I've pushed a new version that moves the code out of the standard qp loop and reduce the number of reinit.

Reply:
OK, I don't exactly know the performance tradeoffs, but I would imagine looping twice would be faster than doing extra reinits... either way as long as it is working correctly, that is the important thing since if I understand correctly, you will be pushing some further optimizations for this code in the near future.